File: fast_tag_pileup/bulk_pileup.jl
using CSV
using DataFrames
using SparseArrays
using DelimitedFiles
using ArgParse

function pileup(scidx_fn::String, beds_fns::String, chrom_sizes::String, out_fn::String; window::Int=201)
    sizes = Dict{String, Int64}()
    open(chrom_sizes, "r") do f
        while !eof(f)
            line = split(readline(f))
            sizes[line[1]] = parse(Int64, line[2])
        end
    end

    scidx = CSV.read(scidx_fn, DataFrame, delim="\t", comment="#")
    contig = scidx[1, "chrom"]
    start = 1
    chrom_dict = Dict{String, SparseMatrixCSC{Int64, Int64}}()
    n = nrow(scidx)
    for i in 1:n
        if scidx[i, "chrom"] != contig
            idx = scidx[start:i - 1, "index"]
            arr = sparse(vcat(idx, idx), vcat(fill(1, i - start), fill(2, i - start)), vcat(scidx[start:i - 1, "forward"], scidx[start:i - 1, "reverse"]), sizes[contig], 2)
            chrom_dict[contig] = arr
            contig = scidx[i, "chrom"]
            start = i
        end
    end
    idx = scidx[start:n, "index"]
    chrom_dict[contig] = sparse(vcat(idx, idx), vcat(fill(1, n - start + 1), fill(2, n - start + 1)), vcat(scidx[start:n, "forward"], scidx[start:n, "reverse"]), sizes[contig], 2)

    f = open(beds_fns, "r")
    beds = readlines(f)
    close(f)

    pileup_mat = zeros(Int64, (length(beds) * 2, window))
    for i in 1:length(beds) # Not sure why there are race conditions when I try to multithread this
        bed = CSV.read(beds[i], DataFrame, delim="\t", header=["Chromosome", "Start", "End", "Name", "Score", "Strand"])
        strand = Vector{Int64}(undef, nrow(bed))
        @simd for j in 1:nrow(bed)
            strand[j] = convert(Int64, bed[j, "Strand"] == "-")
        end

        for j in 1:nrow(bed)
            if bed[j, "Start"] > 0 && bed[j, "End"] <= sizes[bed[j, "Chromosome"]]
                strand_idx = strand[j]
                chrom_sparse = chrom_dict[bed[j, "Chromosome"]]
                start = bed[j, "Start"] - 1
                @simd for k in bed[j, "Start"]:bed[j, "End"]
                    pileup_mat[2 * i + strand_idx - 1, k - start] += chrom_sparse[k, strand_idx + 1]
                end
            end
        end
    end

    writedlm(out_fn, pileup_mat)
end

function main()
    s = ArgParseSettings()
    @add_arg_table s begin
        "--scidx-fn", "-s"
            help = "scidx tab file for sample"
            required = true
        "--beds-fns", "-b"
            help = "file containing list of bed filenames to pileup on"
            required = true
        "--chrom-sizes", "-c"
            help = "chromosome sizes file"
            required = true
        "--out", "-o"
            help = "output file name"
            default = "bulk_pileup.txt"
        "--window", "-w"
            help = "window size in bp"
            arg_type = Int
            default = 201
    end

    args = parse_args(s)

    pileup(args["scidx-fn"], args["beds-fns"], args["chrom-sizes"], args["out"], window=args["window"])
end

main()

File: fast_tag_pileup/bulk_pileup.py
import numpy as np
import os
import re
import h5py
from filelock import SoftFileLock
import argparse

def pileup(scidx_fn: str, bed_fns: list, chrom_sizes: str, out: str, control: bool = False,
           ref_pattern: re.Pattern = re.compile(r'(.+)_(\d+)bp.bed'),
           sample_pattern: re.Pattern = re.compile(r'(\d+)_(.+?)_i5006_BY4741_-_YPD_(.+?)_XO_FilteredBAM')):
    """
    scidx_fn: str
        Path to the scidx file
    bed_fns: list
        List of paths to bed files
    chrom_sizes: str
        Path to the chrom.sizes file
    out: str
        Path to the output h5 file
    control: bool
        If True, the output will be formatted as a control pileup
    ref_pattern: re.Pattern
        Regular expression pattern to extract the reference name from the bed file name
    sample_pattern: re.Pattern
        Regular expression pattern to extract the replicate, target, and condition from the scidx file name
    """
    sizes = {}
    chrom_dict = {}
    with open(chrom_sizes) as f:
        for line in f:
            chrom, size = line.strip().split()
            sizes[chrom] = int(size)
            chrom_dict[chrom] = np.zeros((sizes[chrom], 2), dtype=np.int32)
    
    with open(scidx_fn, 'r') as f:
        contig = None
        f.readline()
        f.readline()
        for line in f:
            line = line.strip().split('\t')
            contig = line[0]
            pos = int(line[1]) - 1
            chrom_dict[contig][pos, 0] = int(line[2])
            chrom_dict[contig][pos, 1] = int(line[3])
    
    sample_id, target, cond = sample_pattern.match(os.path.basename(scidx_fn)).groups()
    lock = SoftFileLock('{}.lock'.format(out))
    with lock:
        with h5py.File(out, 'a') as h5:
            target_cond = '{}-{}'.format(target, cond)
            if h5.get(target_cond) is None:
                h5.create_group(target_cond)
            target_group = h5[target_cond]
            if target_group.get(sample_id) is None:
                target_group.create_group(sample_id)
            
            if control:
                if h5.get('controls') is None:
                    h5.create_group('controls')
                controls = set(h5['controls'][:])
                controls.add(target_cond)
                h5['controls'] = list(controls)

    for bed_fn in bed_fns:
        print(bed_fn)
        ref_name, length = ref_pattern.match(os.path.basename(bed_fn)).groups()
        length = int(length)
        forward_comp = np.zeros(length, dtype=np.int32)
        reverse_comp = np.zeros(length, dtype=np.int32)
        with open(bed_fn, 'r') as f:
            for line in f:
                line = line.strip().split('\t')
                contig = line[0]
                start = int(line[1])
                end = int(line[2])
                if start < 1 or end > sizes[contig] - 1:
                    print('Skipping {}:{}-{} (out of bounds)'.format(contig, start, end))
                    continue
                if line[5] == '-':
                    forward_comp += chrom_dict[contig][end - 1:start - 1:-1, 1]
                    reverse_comp += chrom_dict[contig][end - 1:start - 1:-1, 0]
                else:
                    forward_comp += chrom_dict[contig][start:end, 0]
                    reverse_comp += chrom_dict[contig][start:end, 1]
        
        with lock:
            with h5py.File(out, 'a') as h5:
                rep_group = h5[target_cond][sample_id]
                if rep_group.get(ref_name) is None:
                    rep_group.create_group(ref_name)
                ref_group = rep_group[ref_name]
                ref_group['forward'] = forward_comp
                ref_group['reverse'] = reverse_comp

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('scidx_fn', help='scidx file')
    parser.add_argument('bed_fns', help='file containing list of bed files')
    parser.add_argument('chrom_sizes', help='chrom.sizes file')
    parser.add_argument('out', help='output h5 file')
    parser.add_argument('--control', '-c', action='store_true', help='output control pileup')
    parser.add_argument('--ref_pattern', '-rp', default='(.+)_(\d+)bp.bed',
                        help='regular expression pattern to extract the reference name row width from the bed file name')
    parser.add_argument('--sample_pattern', '-sp', default='(\d+)_(.+?)_i5006_BY4741_-_YPD_(.+?)_XO_FilteredBAM',
                        help='regular expression pattern to extract the ID, target, and condition from the scidx file name')
    args = parser.parse_args()

    with open(args.bed_fns) as f:
        bed_fns = f.read().strip().split('\n')
    pileup(args.scidx_fn, bed_fns, args.chrom_sizes, args.out, control=args.control,
           ref_pattern=re.compile(args.ref_pattern), sample_pattern=re.compile(args.sample_pattern))